src/node_buffer.cc

```diff
@@ -205,41 +306,135 @@ Local<Object> Use(Isolate* isolate, char* data, uint32_t length) {
 }
 
 
-Local<Object> Use(Environment* env, char* data, uint32_t length) {
+Local<Object> Use(Environment* env, char* data, size_t length) {
   EscapableHandleScope scope(env->isolate());
 
-  CHECK_LE(length, kMaxLength);
+  if (using_old_buffer) {
+    CHECK_LE(length, kMaxLength);
 
-  Local<Value> arg = Uint32::NewFromUnsigned(env->isolate(), length);
-  Local<Object> obj = env->buffer_constructor_function()->NewInstance(1, &arg);
+    Local<Value> arg = Uint32::NewFromUnsigned(env->isolate(), length);
+    Local<Object> obj =
+        env->buffer_constructor_function()->NewInstance(1, &arg);
 
-  smalloc::Alloc(env, obj, data, length);
+    smalloc::Alloc(env, obj, data, length);
 
-  return scope.Escape(obj);
+    return scope.Escape(obj);
+  }
+
+  if (length > 0) {
+    CHECK_NE(data, nullptr);
+    CHECK(v8::internal::Internals::IsValidSmi(length));
+  }
+
+  Local<ArrayBuffer> ab =
+      ArrayBuffer::New(env->isolate(),
+                       data,
+                       length,
+                       ArrayBufferCreationMode::kInternalized);
+  Local<Uint8Array> ui = Uint8Array::New(ab, 0, length);
+  ui->SetPrototype(env->buffer_prototype_object());
+  return scope.Escape(ui);
+}
+
+
+void Create(const FunctionCallbackInfo<Value>& args) {
+  Isolate* isolate = args.GetIsolate();
+  Environment* env = Environment::GetCurrent(isolate);
+
+  CHECK(args[0]->IsNumber());
+
+  size_t length = args[0]->IntegerValue();
```

Comment:
Sorry, what I mean is that values that don't fit in a 32 bits unsigned int will wrap around and that's probably not the expected behavior.

I'm not sure if there is an entirely fool-proof way to handle that that doesn't involve knowing the range of a SMI.  `IsValidSmi()` takes an intptr_t and that's going to be 32 bits on 32 bits architectures.

Maybe you should write your own version that operates on an int64_t, just to play it safe:

```c++
inline bool IsValidSmi(uint64_t value) {
  if (sizeof(int32_t) == sizeof(intptr_t)) {
    return value >= -0x40000000 && value <= 0x3fffffff;
  } else {
    return value >= -0x80000000 && value <= 0x7fffffff;
  }
}
```